graphql/core/validation/rules.py

```diff
@@ -234,7 +234,52 @@ def unused_fragment_message(fragment_name):
 
 
 class PossibleFragmentSpreads(ValidationRule):
-    pass
+    def enter_InlineFragment(self, node, *args):
+        frag_type = self.context.get_type()
+        parent_type = self.context.get_parent_type()
+        if frag_type and parent_type and not self.do_types_overlap(frag_type, parent_type):
+            return GraphQLError(
+                self.type_incompatible_anon_spread_message(parent_type, frag_type),
+                [node]
+            )
+
+    def enter_FragmentSpread(self, node, *args):
+        frag_name = node.name.value
+        frag_type = self.get_fragment_type(self.context, frag_name)
+        parent_type = self.context.get_parent_type()
+        if frag_type and parent_type and not self.do_types_overlap(frag_type, parent_type):
+            return GraphQLError(
+                self.type_incompatible_spread_message(frag_name, parent_type, frag_type),
+                [node]
+            )
+
+    @staticmethod
+    def get_fragment_type(context, name):
+        frag = context.get_fragment(name)
+        return frag and type_from_ast(context.get_schema(), frag.type_condition)
+
+    @staticmethod
+    def do_types_overlap(t1, t2):
+        if t1 == t2:
+            return True
+        if isinstance(t1, GraphQLObjectType):
+            if isinstance(t2, GraphQLObjectType):
+                return False
+            return t2.get_possible_types().index(t1) != -1
```

Comment:
`list.index` throws if there's no element found. Change to `t1 in t2.get_possible_types()`?